File: javascript/test/CucumberExpressionTokenizerTest.ts
import assert from 'assert'
import fs from 'fs'
import glob from 'glob'
import yaml from 'js-yaml'

import { describe, it } from 'minispec'

import CucumberExpressionError from '../src/CucumberExpressionError.js'
import CucumberExpressionTokenizer from '../src/CucumberExpressionTokenizer.js'
import { testDataDir } from './testDataDir.js'

type Expectation = {
  expression: string
  expected_tokens?: unknown
  exception?: string
}

describe('CucumberExpressionTokenizer', async () => {
  for (const path of glob.sync(`${testDataDir}/cucumber-expression/tokenizer/*.yaml`)) {
    const expectation = yaml.load(fs.readFileSync(path, 'utf-8')) as Expectation
    it(`tokenizes ${path}`, async () => {
      const tokenizer = new CucumberExpressionTokenizer()
      if (expectation.expected_tokens !== undefined) {
        const tokens = tokenizer.tokenize(expectation.expression)
        assert.deepStrictEqual(
          JSON.parse(JSON.stringify(tokens)), // Removes type information.
          expectation.expected_tokens
        )
      } else if (expectation.exception !== undefined) {
        assert.throws(() => {
          tokenizer.tokenize(expectation.expression)
        }, new CucumberExpressionError(expectation.exception))
      } else {
        throw new Error(
          `Expectation must have expected_tokens or exception: ${JSON.stringify(expectation)}`
        )
      }
    })
  }
})

File: dart/test/cucumber_expression_tokenizer_test.dart
import 'package:cucumber_expressions/src/cucumber_expression_tokenizer.dart';
import 'package:cucumber_expressions/src/errors.dart';
import 'package:test/test.dart';
import 'package:yaml/yaml.dart';

import 'test_data_dir.dart';

void main() {
  group('CucumberExpressionTokenizer', () {
    for (final file
        in yamlFilesIn('$testDataDir/cucumber-expression/tokenizer')) {
      final expectation = loadYaml(file.readAsStringSync()) as YamlMap;
      test('tokenizes ${file.path}', () {
        final expression = expectation['expression'] as String;
        if (expectation['expected_tokens'] != null) {
          final tokens = CucumberExpressionTokenizer().tokenize(expression);
          final actual = tokens
              .map(
                (t) => {
                  'type': t.type.value,
                  'start': t.start,
                  'end': t.end,
                  'text': t.text,
                },
              )
              .toList();
          final expected = (expectation['expected_tokens'] as YamlList)
              .cast<YamlMap>()
              .map(
                (t) => {
                  'type': t['type'],
                  'start': t['start'],
                  'end': t['end'],
                  'text': t['text'],
                },
              )
              .toList();
          expect(actual, equals(expected));
        } else if (expectation['exception'] != null) {
          expect(
            () => CucumberExpressionTokenizer().tokenize(expression),
            throwsA(
              isA<CucumberExpressionException>().having(
                (e) => e.message,
                'message',
                equals(expectation['exception']),
              ),
            ),
          );
        } else {
          fail('Expectation must have expected_tokens or exception');
        }
      });
    }
  });
}
